File: src/web/magmac/app/lang/node/StructureStatement.ts
import { Option } from "../../../../magmac/api/Option";
import { Lists } from "../../../../magmac/api/collect/list/Lists";
import { CompileResult } from "../../../../magmac/app/compile/error/CompileResult";
import { MapNode } from "../../../../magmac/app/compile/node/MapNode";
import { Node } from "../../../../magmac/app/compile/node/Node";
import { NodeRule } from "../../../../magmac/app/compile/rule/NodeRule";
import { OrRule } from "../../../../magmac/app/compile/rule/OrRule";
import { Rule } from "../../../../magmac/app/compile/rule/Rule";
import { StripRule } from "../../../../magmac/app/compile/rule/StripRule";
import { SuffixRule } from "../../../../magmac/app/compile/rule/SuffixRule";
import { TypeRule } from "../../../../magmac/app/compile/rule/TypeRule";
import { Deserializers } from "../../../../magmac/app/lang/Deserializers";
import { LazyRule } from "../../../../magmac/app/lang/LazyRule";
export class StructureStatement {
	deserialize(node : Node) : Option<CompileResult<JavaStructureMember>>;
	createStructureStatementRule(definitionRule : Rule, valueRule : LazyRule) : Rule;
}

File: src/java/magmac/app/lang/node/StructureStatement.java
package magmac.app.lang.node;

import magmac.api.Option;
import magmac.api.collect.list.Lists;
import magmac.app.compile.error.CompileResult;
import magmac.app.compile.node.MapNode;
import magmac.app.compile.node.Node;
import magmac.app.compile.rule.NodeRule;
import magmac.app.compile.rule.OrRule;
import magmac.app.compile.rule.Rule;
import magmac.app.compile.rule.StripRule;
import magmac.app.compile.rule.SuffixRule;
import magmac.app.compile.rule.TypeRule;
import magmac.app.lang.Deserializers;
import magmac.app.lang.LazyRule;

public record StructureStatement() implements JavaStructureMember {
    public static Option<CompileResult<JavaStructureMember>> deserialize(Node node) {
        return Deserializers.deserializeWithType(node, "structure-statement").map(deserializer -> deserializer.complete(StructureStatement::new));
    }

    public static Rule createStructureStatementRule(Rule definitionRule, LazyRule valueRule) {
        Rule definition = new NodeRule("value", new OrRule(Lists.of(
                definitionRule,
                AssignmentNode.createAssignmentRule(definitionRule, valueRule))
        ));

        return new TypeRule("structure-statement", new StripRule(new SuffixRule(definition, ";")));
    }
}
